File: 프로그래머스/2/42587. 프로세스/프로세스.js
class Queue {
    constructor() {
        this.queue = [];
    }
    
    enqueue(n) {
        this.queue.push(n);
    }
    
    dequeue() {
        return this.queue.shift();
    }
    
    isEmpty() {
        return this.queue.length === 0;
    }
}

function solution(priorities, location) {
    let answer = 0;
    
    const queue = new Queue();
    for(let i = 0; i < priorities.length; i++) {
        queue.enqueue({
            idx: i,
            priority: priorities[i],
        });
    }
    
    let cnt = 0;
    
    while(!queue.isEmpty()) {
        const process = queue.dequeue();
        
        let flag = false;
        
        for(let wait of queue.queue) {
            if(process.priority < wait.priority) {
                flag = true;
                break;
            }
        }
        
        if(flag) {
            queue.enqueue(process);
            continue;
        }
        
        cnt++;
        
        if(process.idx === location) {
            answer = cnt;
            break;
        }
    }
    
    return answer;
}

File: 프로그래머스/2/42587. 프로세스/프로세스.java
import java.util.*;

class Solution {
    public int solution(int[] priorities, int location) {
        int N = priorities.length;
        int[] orders = new int[priorities.length];
        
        Queue<int[]> waitingQueue = new LinkedList<>();
        for(int i = 0; i < N; i++) {
            waitingQueue.add(new int[]{i, priorities[i]});
        }
        
        int rank = 1;
        while(!waitingQueue.isEmpty()) {
            int[] arr = waitingQueue.poll();
            int idx = arr[0];
            int p = arr[1];
            
            Queue<int[]> temp = new LinkedList<>();
            boolean flag = true;
            while(!waitingQueue.isEmpty()) {
                int[] temp_arr = waitingQueue.poll();
                int x = temp_arr[1];
                if(x > p) {
                    flag = false;
                }
                temp.add(temp_arr);
            }
            
            while(!temp.isEmpty()) {
                waitingQueue.add(temp.poll());
            }
            
            if(flag) {
                orders[idx] = rank++;
            } else {
                waitingQueue.add(arr);
            }
        }
        
        return orders[location];
    }
}